### ISSUE TITLE
2.7 Bidirectional Relationship field isn't updated automatically to the opening modal when adding related record - EXTEND for more options

### ISSUE THREAD

**Issue by renaudham** [pods-framework/pods#4269]
Extending this
https://github.com/pods-framework/pods/issues/4249#issuecomment-304528310

I think we need to go further with more choices and options

I copy some previous posts.
------------
Hi
This, I think, a very important element of this new feature, in term of UI and UX, to make it usable by most users.

**First lets note that it can be working ONLY on EDIT form mode. (on Add new, of course no ID exist yet to refer to)

**What is obviously needed is that
When we open the modal related form
the related field should appear automatically set up with the correct value.
Once loaded, so the value selected would be visible to the user, on this field - no on saving.

**To make it understandable let me give a use case

A project form, in EDIT mode, with multiples bugs elements as a related field list view.
And a Add new button.

On add, modal window with a Bug form, and a Project single selector is there
and needs to be auto set to the correct project ID.

**My own guess as I feel is that
*We add parameters in the url sent by JS
wp-admin/post-new.php?post_type=bugs&pods_modal=1&rel_id=1&target_relation=project
rel_id = the parent Project id (coming form the related field data-item-id)
target_relation=the field name

*Then in the relation field code,when we display the proejct field in the bug form, we only and simply need to detected if there is a param target_relation with the same name.
and catch the rel_id ID to set as default.

In case of ajax this should be set in JS after loading of course, but its not a big deal at all

If you don't plan to do it,
I will modify the pick.php and related JS myself for my own, as I really need this.

Saddly all those codes are missing of hook actions and/or filters so we could add this out of the core.
Any plan to add more hooks in the code?
thanks

----------------

jimtrue commented 2 days ago
Add New 'relationship' if you're allowed to do a multiple select will always be available, because you might be adding new 'child record'. When you're in the Child Record 'modal', the Add New is not available or that would take you into an 'inception' style interface. We only allow adding child records from the parent. Even if you are Editing the current record, you might need to Add a New Child record.
 @renaudham
     
renaudham commented 2 days ago
Hi jimtrue.
Of course, I agree that on child modal form they always need to be hidden.
What I was meaning in my last point was about allowing in the main parent form
to show it but only in case of an EDIT form, and not when we are doing a Add new form.

Its already an issue in the current situation of your code, without changing anything.:

You add a new entity Project
in this one you have a field multiple bugs, with a ADD NEW Button
If you do Add new bug before saving your current project
then you will have a modal with a bug form and a Project selector.
But in this project selector you have of course ONLY old project pre existing.
And then the only thing you cuold do is add a new bug to on of the old other projects, with no relaionship with the current new object you are building.

This is a kind of usability non sens.

So we should allow a setting to show this button only in case of EDIT,
I believe that this is a usability need.

-------------

renaudham commented 2 days ago • edited
For info I managed very quickly to add the first part I was talking about

to have this url auto generated for the modal
wp-admin/post-new.php?post_type=bugs&pods_modal=1&rel_id=9&target_relation=project

just adding this

//ADDED BY REN 20170528
self::$type . '_set_related_auto' => array(
'label' => __('Add New > Auto set child related field', 'pods'),
'help' => __('On adding a new record, the related child field will be auto set to the value of the current parent', 'pods'),

            /*   'depends-on' => array(//array( self::$type . '_object' => 'user' )
                self::$type . '_allow_add_new'   => 1,
            ),

            'excludes-on' => array(
                self::$type . '_allow_add_new'   => false,
            ),*/
            'type' => 'boolean',
            'default' => 1,
        ),
        self::$type . '_related_auto_child_slug' => array(
            'label' => __('Add New > child related field slug name', 'pods'),
            'help' => __('On adding a new record, define the related child field slug name', 'pods'),
            'default' => '',
            'type' => 'text',
        ),
public function build_dfv_field_config($args)
{
.....
$iframe_src = '';

    // Potential valid modal target if we've set the file name
    if (!empty($file_name)) {
        // We extend wp.media.view.Modal for modal add/edit, we must ensure we load the template for it
        wp_enqueue_media();

        // @todo: Replace string literal with defined constant
        $query_args['pods_modal'] = 1;
//ADDED BY REN 20170528
if($args->options['pick_set_related_auto']==1){
$query_args['rel_id'] = $_GET['id'];
$query_args['target_relation'] = $args->options['pick_related_auto_child_slug'];//$args->options['name']; pick_set_related_auto pick_related_auto_child_slug

        }
     

        // $query_args['toto'] = 1;

        // Add args we always need
        $iframe_src = add_query_arg($query_args, admin_url($file_name));
    }
Note that the 2 extra config are not yet hiding/showing in case we use or not Add New button

But I think its the simple way to go
I will make the second part and share my idea, if usefull for anybody

**Comment by jimtrue** [comment]
@renaudham You duplicated so much of my original issue that is was hard to follow what you were actually requesting here. #4249 Covers the details of setting the bidirectional field in the child record to the value of the parent (opener) window.

What kind of controls/capabilities, specifically, are you looking for in reference to the Child Records? At present we have Add New and Edit

It sounds like you're wanting to restrict certain actions when the parent post (the one opening the modal) is New itself. Or you want a filter to be able to restrict those yourself. Where would you like the filter to be located, specifically to provide you the most benefit?

**Comment by renaudham** [comment]
Hi.

Ok i modified it by myself anyway by changing some of the core code....
I will wait the future version to see the evolution, if its doing what I need or not.

Regards

**Comment by quasel** [comment]
Hey @renaudham reading this again "by solved the first part this was about #4249 an if yes could you provide a PR to look at for @pglewis - because that's on our ToDo list! 

Still have no idea what you meant  by
> I think we need to go further with more choices and options

but if you have any ideas improving things around modal / DFV / that's a planned part for 2.8 ! 
any help appreciated!

**Comment by renaudham** [comment]
Hi

Ok so I should show screenshot and code excerpt from what I have done and
explain.

Where should I share this, here?

Thanks



*Renaud Hamelin*
Dirigeant, CTO, et Lead développeur
Blue Origami Digital
(SARL, Société en création)
Studio solutions web et solutions cognitives
Conception, Consulting et Production


*Strasbourg - Paris*Skype : renaudham69
Phone : 06 37 67 46 46
Linkedin: https://fr.linkedin.com/in/renaudhamelin

**Comment by quasel** [comment]
Best to do it in #4249 - if it's about profiling related items from bidirectional :D - and best thing is always to make a  Pull Request if possible ;)
